Index: crc.cpp
===================================================================
--- crc.cpp
+++ crc.cpp
@@ -30,7 +30,7 @@
     CRCTab[I]=crc_tables[0][I]=C;
   }
 
-	for (uint I=0;I<=256;I++) // Build additional lookup tables.
+	for (uint I=0;I<256;I++) // Build additional lookup tables.
   {
 		uint C=crc_tables[0][I];
 		for (uint J=1;J<8;J++)
@@ -49,7 +49,7 @@
   byte *Data=(byte *)Addr;
 
   // Align Data to 8 for better performance.
-  for (;Size>0 && ((long)Data & 7);Size--,Data++)
+  for (;Size>0 && (reinterpret_cast<size_t>(Data) & 7);Size--,Data++)
     StartCRC=crc_tables[0][(byte)(StartCRC^Data[0])]^(StartCRC>>8);
 
   for (;Size>=8;Size-=8,Data+=8)
